manifests/init.pp

```diff
@@ -1,160 +1,93 @@
+  Hash                             $ds_config                    = {
+    'root_dn'        => "cn=${pki_security_domain} Directory Manager",
+    'base_dn'        => sprintf('ou=%s,%s',$pki_security_domain, simplib::ldap::domain_to_dn($facts['domain'], false)),
+    'admin_password' => simplib::passgen('389-ds-simp-pki', {'length' => 64, 'complexity' => 0 })
+  },
+  Hash                             $cas                          = {
     'simp-pki-root' => {
-      'root_ca'            => true,
-      'enable_kra'         => false,
+      'root_ca'             => true,
+      'pki_security_domain' => $pki_security_domain,
+      # The following two items need to be here so that the other CA instances
+      # can use them for connecting to the security domain
+      'admin_user'         => 'caadmin',
+      'admin_password'     => simplib::passgen("${pki_security_domain}_simp-pki-root", { 'length' => 64, 'complexity' => 0 }),
       'http_port'          => 4508,
       'https_port'         => 4509,
       'tomcat_ajp_port'    => 4506,
       'tomcat_server_port' => 4507
     },
     'simp-puppet-pki' => {
-      'enable_scep'        => true,
-      'http_port'          => 5508,
-      'https_port'         => 5509,
-      'tomcat_ajp_port'    => 5506,
-      'tomcat_server_port' => 5507,
-      'parent_ca'          => 'simp-pki-root'
+      'enable_kra'          => true,
+      'enable_scep'         => true,
+      'http_port'           => 5508,
+      'https_port'          => 5509,
+      'tomcat_ajp_port'     => 5506,
+      'tomcat_server_port'  => 5507,
+      'parent_ca'           => 'simp-pki-root'
     },
     'simp-site-pki' => {
-      'enable_scep'        => true,
-      'http_port'          => 8080,
-      'https_port'         => 8443,
-      'tomcat_ajp_port'    => 8440,
-      'tomcat_server_port' => 8441,
-      'parent_ca'          => 'simp-pki-root'
+      'enable_kra'          => true,
+      'enable_scep'         => true,
+      'http_port'           => 8080,
+      'https_port'          => 8443,
+      'tomcat_ajp_port'     => 8440,
+      'tomcat_server_port'  => 8441,
+      'parent_ca'           => 'simp-pki-root'
     }
-  }
+  },
+  Hash                             $custom_cas                   = {},
+  Boolean                          $enable_haveged               = true
 ){
-  include haveged
+  if $enable_haveged { include haveged }
   include simp_pki_service::config
 
-  simp_pki_service::directory_server { 'simp-pki-ds':
-    dirsrv_base_dn       => $dirsrv_base_dn,
-    dirsrv_root_dn       => $dirsrv_root_dn,
-    admin_password       => $admin_password,
-    enable_admin_service => $dirsrv_enable_admin_service,
-    listen_address       => $dirsrv_listen_address
-  }
+  simp_pki_service::directory_server { 'simp-pki-ds': * => $ds_config }
 
   $_cas = deep_merge($cas, $custom_cas)
 
   keys($_cas).each |String $ca_id| {
     if $_cas[$ca_id]['root_ca'] {
       simp_pki_service::ca { $ca_id:
```

Comment:
@lnemsick-simp Hmmm...technically, this will blow up as soon as it tries to hook things together in the underlying CA class. Do you think this is sufficient or should we yell more loudly?

Reply:
If the error messages are clear, I don't think extra validation is warranted.  However, if what spews out is some obscure puppet/ruby error, then I would rather validate and tell the user clearly what the problem is.  This could be a future enhancement.

Reply:
Created [SIMP-4427](https://simp-project.atlassian.net/browse/SIMP-4427) for this